File: ios/Cove/Flows/TapSignerFlow/TapSignerSetupRetryView.swift
//
//  TapSignerSetupRetryView.swift
//  Cove
//
//  Created by Praveen Perera on 3/25/25.
//

import CoveCore
import SwiftUI
import UniformTypeIdentifiers

struct TapSignerSetupRetry: View {
    @Environment(AppManager.self) private var app
    @Environment(TapSignerManager.self) private var manager

    let tapSigner: TapSigner
    let response: SetupCmdResponse
    @State private var isSubmitting = false

    var body: some View {
        TapSignerAdaptiveLayout { usesFlexibleSpacing in
            TapSignerContinuationContent(
                usesFlexibleSpacing: usesFlexibleSpacing,
                title: "Setup in Progress",
                message: "Your TAPSIGNER saved its setup progress. Continue to finish setup.",
                isSubmitting: isSubmitting,
                cancelAction: cancel,
                continueAction: continueSetup
            )
        }
        .background(TapSignerResultBackground())
        .scrollIndicators(.hidden)
        .navigationBarHidden(true)
    }

    private func cancel() {
        app.sheetState = nil
    }

    private func continueSetup() {
        guard !isSubmitting else { return }

        isSubmitting = true

        Task {
            let nfc = manager.getOrCreateNfc(tapSigner)
            let result = await nfc.continueSetup(response)

            await MainActor.run {
                isSubmitting = false

                switch result {
                case let .success(.complete(complete)):
                    manager.resetRoute(to: .setupSuccess(tapSigner, complete))
                case let .success(.retry(next)):
                    manager.resetRoute(to: .setupRetry(tapSigner, .retry(next)))
                case .failure:
                    app.sheetState = nil
                    app.alertState = .init(
                        .tapSignerSetupFailed(
                            message: "TapSigner setup failed. Please try again."
                        )
                    )
                }
            }
        }
    }
}

#Preview {
    TapSignerContainer(
        route:
        .setupRetry(
            tapSignerPreviewNew(preview: true),
            tapSignerSetupRetryContinueCmd(preview: true)
        )
    )
    .environment(AppManager.shared)
}

File: android/app/src/main/java/org/bitcoinppl/cove/flows/TapSignerFlow/TapSignerSetupRetryView.kt
package org.bitcoinppl.cove.flows.TapSignerFlow

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlin.coroutines.cancellation.CancellationException
import kotlinx.coroutines.launch
import org.bitcoinppl.cove.AppManager
import org.bitcoinppl.cove.TaggedItem
import org.bitcoinppl.cove_core.AppAlertState
import org.bitcoinppl.cove_core.SetupCmdResponse
import org.bitcoinppl.cove_core.TapSignerRoute

/** Retry an opaque setup continuation returned by Rust. */
@Composable
fun TapSignerSetupRetryView(
    app: AppManager,
    manager: TapSignerManager,
    tapSigner: org.bitcoinppl.cove_core.tapcard.TapSigner,
    response: SetupCmdResponse,
    modifier: Modifier = Modifier,
) {
    val scope = rememberCoroutineScope()
    var isSubmitting by remember { mutableStateOf(false) }

    Column(
        modifier = modifier.fillMaxSize().padding(horizontal = 16.dp),
        verticalArrangement = Arrangement.SpaceBetween,
    ) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(top = 20.dp),
            horizontalArrangement = Arrangement.Start,
        ) {
            TextButton(onClick = { app.sheetState = null }) {
                Text("Cancel", fontWeight = FontWeight.SemiBold)
            }
        }

        Column(
            modifier = Modifier.fillMaxWidth().weight(1f),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center,
        ) {
            Icon(
                imageVector = Icons.Default.Refresh,
                contentDescription = null,
                modifier = Modifier.size(100.dp),
                tint = MaterialTheme.colorScheme.primary,
            )

            Spacer(modifier = Modifier.height(20.dp))

            Text(
                text = "Setup in Progress",
                style = MaterialTheme.typography.headlineMedium,
                fontWeight = FontWeight.Bold,
            )

            Text(
                text = "Your TAPSIGNER saved its setup progress. Continue to finish setup.",
                style = MaterialTheme.typography.bodyMedium,
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(top = 12.dp),
            )
        }

        Button(
            onClick = {
                if (isSubmitting) return@Button
                isSubmitting = true
                scope.launch {
                    try {
                        val nfc = manager.getOrCreateNfc(tapSigner)
                        manager.beginScan("Hold your phone near the TapSigner to continue setup")

                        val nextResponse =
                            nfc.continueSetup(response, manager.operationCallbacks())
                        when (nextResponse) {
                            is SetupCmdResponse.Complete -> {
                                manager.resetRoute(
                                    TapSignerRoute.SetupSuccess(tapSigner, nextResponse.v1),
                                )
                            }
                            is SetupCmdResponse.Retry -> {
                                manager.resetRoute(TapSignerRoute.SetupRetry(tapSigner, nextResponse))
                            }
                        }
                    } catch (error: CancellationException) {
                        throw error
                    } catch (_: Exception) {
                        app.sheetState = null
                        app.alertState =
                            TaggedItem(
                                AppAlertState.TapSignerSetupFailed(
                                    "TapSigner setup failed. Please try again.",
                                ),
                            )
                    } finally {
                        manager.endScan()
                        isSubmitting = false
                    }
                }
            },
            enabled = !isSubmitting,
            modifier = Modifier.fillMaxWidth().padding(bottom = 30.dp).testTag("tapSignerSetupRetry.continue"),
        ) {
            Text(if (isSubmitting) "Working…" else "Continue Setup")
        }
    }
}
